Update the locator for the region select

bobsilverberg opened the pull request:
Change tap() to send_keys() as tap() no longer works

This is discussed in issue #87. We may have to accept this workaround for now, in order to restore coverage, but it deserves more investigation.

@davehunt r?

bobsilverberg commented:
I added a second commit to this PR as I was trying to get the create/confirm PIN test passing, and fixing the debug was only part of that. We likely also need the added sleep to get the test passing reliably, until we can find a better solution.

bobsilverberg commented:
This failure can be reproduced by running https://github.com/mozilla/marketplace-tests-gaia/blob/master/marketplacetests/tests/test_marketplace_create_confirm_pin.py

davehunt commented on marketplacetests/marketplace/regions/debug.py:
```diff
@@ -21,6 +21,6 @@ def tap_back(self):
 
     def select_region(self, region):
-        self.marionette.find_element(*self._region_select_locator).tap()
+        self.marionette.find_element(*self._region_select_locator).send_keys(region)
```
I believe the issue here is that Marionette is not scrolling the element fully into view. This was fixed in [bug 937053](https://bugzilla.mozilla.org/show_bug.cgi?id=937053), which landed after the v2.0 branch was created. Perhaps a better workaround would be to force the element into view using:

```python
element = self.marionette.find_element(*self._region_select_locator)
self.marionette.execute_script('arguments[0].scrollIntoView(false);', [element])
element.tap()
```

I've tested this locally with success.

davehunt replied:
Also the full viewport screenshots landed after the v2.0 branch was created, otherwise this might have been easier to debug (it would show the element was mostly off-screen).

davehunt commented on marketplacetests/payment/app.py:
```diff
@@ -82,6 +83,9 @@ def tap_cancel_pin(self):
         Wait(marionette=self.marionette).until(lambda m: button.is_enabled())
+        # TODO: Even with the above wait, marionette often fails to tap on
+        # the button successfully. Adding a sleep fixes this for now.
+        time.sleep(1)
```
I'm trying to replicate the failure that this addresses but so far without success. I have replicated an issue where the keyboard is not dismissed after entering the password, I think this is related to #86 so will comment there.

davehunt replied:
If it's possible that the button is partially outside of the viewport then you could try the `scrollIntoView(false)` workaround here.

bobsilverberg commented:
Thanks @davehunt, `scrollIntoView` did indeed do the trick! I have updated this PR with that suggestion, and removed the other code that deals with the PIN input, so that this commit is only about the debug screen and region select. Could I bother you for a final r? on the code?

davehunt commented on marketplacetests/marketplace/regions/debug.py:
```diff
@@ -21,6 +21,8 @@ def tap_back(self):
     def select_region(self, region):
-        self.marionette.find_element(*self._region_select_locator).tap()
+        element = self.marionette.find_element(*self._region_select_locator)
+        self.marionette.execute_script('arguments[0].scrollIntoView(false);', [element])
```
Could we add a comment mentioning the bug? If in the future we don't need to worry about v2.0 we should be able to remove this workaround. Something like `# Bug 937053 - tap() method should calculate elementInView from the coordinates of the tap`

With that change, r+

bobsilverberg replied:
Good idea. Should we be even more explicit though? That comment just mentions a bug, but not the fact that we won't need this workaround in later gaia versions. So maybe the comment should mention that. What do you think?

davehunt replied:
Yeah, that sounds good to me.

bobsilverberg commented:
Thanks for the review, @davehunt. I added the comment and am going to merge this.
